allennlp/interpret/attackers/input_reduction.py

```diff
@@ -48,36 +54,57 @@ def attack_from_json(self, inputs: JsonDict = None,
-                        outputs[output] = outputs[output][0]
-
-                # Check if any fields have changed, if so, break loop
-                if "tags" not in current_instance:
-                    if any(current_instance[field] != fields_to_compare[field] for field in fields_to_compare):
-                        break
-
-                # special case for sentence tagging (we have tested NER)
-                else:
-                    if smallest_idx != -1:
-                        del tag_mask[smallest_idx]
-                    cur_tags = [outputs["tags"][x] for x in range(len(outputs["tags"])) if tag_mask[x]]
-                    if cur_tags != original_tags:
-                        break
-
-                # remove a token from the input
-                current_tokens = deepcopy(current_text_field.tokens)
-                current_instance, smallest_idx = _remove_one_token(current_instance,
-                                                                   input_field_to_attack,
-                                                                   grads[grad_input_field],
-                                                                   ignore_tokens)
+            while len(current_tokens) > num_ignore_tokens and current_candidates:
+                # sort current candidates by smallest length (we want to remove as many tokens as possible)
+                def get_length(input_instance: Instance):
+                    input_text_field: TextField = input_instance[input_field_to_attack]  # type: ignore
+                    return len(input_text_field.tokens)
+                current_candidates = heapq.nsmallest(self.beam_size,
+                                                     current_candidates,
+                                                     key=lambda x: get_length(x[0]))
+
+                beam_candidates = deepcopy(current_candidates)
+                current_candidates = []
+                for beam_instance, smallest_idx in beam_candidates:
+                    # get gradients and predictions
+                    grads, outputs = self.predictor.get_gradients([beam_instance])
+
+                    for output in outputs:
+                        if isinstance(outputs[output], torch.Tensor):
+                            outputs[output] = outputs[output].detach().cpu().numpy().squeeze().squeeze()
+                        elif isinstance(outputs[output], list):
+                            outputs[output] = outputs[output][0]
+
+                    # Check if any fields have changed, if so, next beam
+                    if "tags" not in current_instance:
+                        # relabel beam_instance since last iteration removed an input token
+                        beam_instance = self.predictor.predictions_to_labeled_instances(beam_instance, outputs)[0]
+                        if any(beam_instance[field] != fields_to_compare[field] for field in fields_to_compare):
+                            continue
```

Comment:
This method is pretty nested and long.  Can you break it up into several methods?  E.g., an `attack_instance()` method that gets called in the top-level for loop at line 40, and maybe a `beam_step` method, or something.  And you could pull out this check into a method like `has_changed` (and that method could potentially be shared with hotflip).

Reply:
Ok i added a those three suggestions, and shared the has_changed with hotflip

Reply:
actually, I don't think the `beam_step` function is necessary, it has 8 arguments and makes things harder to follow in my opinion.

Reply:
That's fine, it was just a suggestion.  Probably `attack_instance` and `has_changed` are sufficient to make this reasonable.

Reply:
Oh, I meant taking this whole outer block, including the special case for tagging.

Reply:
Probably easiest to structure it like this:
```python
beam_instance = self.predictor.predictions_to_labeled_instances(beam_instance, outputs)[0]
if util.instance_has_changed(beam_instance, fields_to_compare):
    continue

current_text_field: TextField ...
```
And the special case for NER lives inside `instance_has_changed`.